hummingbot/connector/derivative/phemex_perpetual/phemex_perpetual_api_order_book_data_source.py

```diff
@@ -0,0 +1,234 @@
+                subscribe_request: WSJSONRequest = WSJSONRequest(payload)
+                async with self._api_factory.throttler.execute_task(limit_id=CONSTANTS.WSS_MESSAGE_LIMIT_ID):
+                    await ws.send(subscribe_request)
+            self.logger().info("Subscribed to public order book, trade and funding info channels...")
+            safe_ensure_future(self.ping_loop(ws=ws))
+        except asyncio.CancelledError:
+            raise
+        except Exception:
+            self.logger().exception("Unexpected error occurred subscribing to order book trading and delta streams...")
+            raise
+
+    def _channel_originating_message(self, event_message: Dict[str, Any]) -> str:
+        channel = ""
+        if self._diff_messages_queue_key in event_message:
+            channel = self._diff_messages_queue_key
+        elif self._trade_messages_queue_key in event_message:
+            channel = self._trade_messages_queue_key
+        elif event_message.get("method", None) == self._funding_info_messages_queue_key:
+            channel = self._funding_info_messages_queue_key
+        elif event_message.get("result", None) == "pong":
+            self.pong_received_event.set()
+        return channel
+
+    async def _parse_order_book_diff_message(self, raw_message: Dict[str, Any], message_queue: asyncio.Queue):
+        timestamp: float = time.time()
+        raw_message["symbol"] = await self._connector.trading_pair_associated_to_exchange_symbol(raw_message["symbol"])
+        if raw_message["type"] == "incremental":
+            book = raw_message[self._diff_messages_queue_key]
+            order_book_message: OrderBookMessage = OrderBookMessage(
+                OrderBookMessageType.DIFF,
+                {
+                    "trading_pair": raw_message["symbol"],
+                    "update_id": raw_message["sequence"],
+                    "bids": book["bids"],
+                    "asks": book["asks"],
+                },
+                timestamp=timestamp,
+            )
+            message_queue.put_nowait(order_book_message)
+
+    async def _parse_trade_message(self, raw_message: Dict[str, Any], message_queue: asyncio.Queue):
+        raw_message["symbol"] = await self._connector.trading_pair_associated_to_exchange_symbol(raw_message["symbol"])
+        for trade in raw_message[self._trade_messages_queue_key]:
+            mapped_trade = TradeStructure(*trade)
+            trade_message: OrderBookMessage = OrderBookMessage(
+                OrderBookMessageType.TRADE,
+                {
+                    "trading_pair": raw_message["symbol"],
+                    "trade_type": mapped_trade.side.upper(),
+                    "trade_id": mapped_trade.timestamp,
+                    "update_id": raw_message["sequence"],
+                    "price": mapped_trade.price,
+                    "amount": mapped_trade.amount,
+                },
+                timestamp=mapped_trade.timestamp // 1e9,
```

Comment:
It is allowed to pass the timestamp as float. If we transform it to int when converting it to seconds, we change the original value and there is no need.

Reply:
done